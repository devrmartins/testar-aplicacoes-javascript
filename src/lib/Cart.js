import { find, remove } from "lodash";
export default class Cart {
    total = 0;
    items = [];

    add(item) {
        const itemToFind = { product: item.product }
        const itemFind = find(this.items, itemToFind);

        if ( itemFind ) {
            itemFind.quantity += item.quantity
        } else {
            this.items.push(item);
        }
        
    }

    getItem(product) {
        return find(this.items, { product })
    }

    getTotalItems() {
        return this.items.length;
    }

    getTotal() {
        return this.items.reduce(
            (acc, item) => acc + item.quantity * item.product.price,
            0
        );
    }

    checkout() {
        return {
            total: this.getTotal(),
            items: this.items
        }
    }

    remove(product) {
        remove(this.items, { product })
    }
}
