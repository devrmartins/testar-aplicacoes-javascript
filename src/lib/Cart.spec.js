import Cart from "./Cart";

describe("Cart", () => {
    let cart;
    let product;
    let product2;

    beforeEach(() => {
        cart = new Cart();
        product = {
            title: "Adidas running shoes - men",
            price: 35388,
        };

        product2 = {
            title: "Adidas running shoes - women",
            price: 42763,
        };
    });

    it("should return 0 when getTotal() is executed in a newly created instancy", () => {
        expect(cart.getTotal()).toEqual(0);
    });

    it("should multiply quantity and price and receive the total amount", () => {
        cart.add({
            product,
            quantity: 2,
        });

        expect(cart.getTotal()).toEqual(70776);
    });

    it("should sum quantity when the product already exists", () => {
        cart.add({
            product,
            quantity: 2,
        });

        cart.add({
            product,
            quantity: 1,
        });

        expect(cart.items[0].quantity).toEqual(3);
        expect(cart.getTotal()).toEqual(106164);
    });

    it("should update total when a product gets included an then removed", () => {
        cart.add({
            product,
            quantity: 3,
        });
        cart.add({
            product: product2,
            quantity: 3,
        });

        cart.remove(product);

        expect(cart.getTotal()).toEqual(128289);
    });

    it("should return the total items on cart", () => {
        cart.add({
            product,
            quantity: 2,
        });

        expect(cart.getTotalItems()).toEqual(1);
    });

    it("should ensure no more than on product exists at a time", () => {
        cart.add({
            product,
            quantity: 2,
        });
        cart.add({
            product,
            quantity: 1,
        });

        expect(cart.getTotalItems()).toEqual(1);
    });

    it("should return a item searching by product", () => {
        let item1 = {
            product,
            quantity: 5,
        };

        let item2 = {
            product: product2,
            quantity: 5,
        };

        cart.add(item1);
        cart.add(item2);

        expect(cart.getItem(product2)).toMatchObject(item2);
    });

    it('should return an object with the total and the list of items', () => {
        cart.add({
            product,
            quantity: 2,
        });

        cart.add({
            product: product2,
            quantity: 5,
        });

        expect(cart.checkout()).toMatchSnapshot()
    })

    it('should return an object with the total and the list of items when summary is called', () => {
        cart.add({
            product,
            quantity: 2,
        });

        cart.add({
            product: product2,
            quantity: 5,
        });

        expect(cart.summary()).toMatchSnapshot()
    })

    it('should reset the cart when checkout() is called', () => {
        cart.add({
            product,
            quantity: 2,
        });

        cart.checkout()

        expect(cart.getTotal()).toEqual(0)
    })
});
