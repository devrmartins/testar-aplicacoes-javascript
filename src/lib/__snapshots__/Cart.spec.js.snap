// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 42763,
        "title": "Adidas running shoes - women",
      },
      "quantity": 5,
    },
  ],
  "total": 284591,
}
`;

exports[`Cart should return an object with the total and the list of items when summary is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 42763,
        "title": "Adidas running shoes - women",
      },
      "quantity": 5,
    },
  ],
  "total": 284591,
}
`;
